docs/design/password-key-derivation.md

```diff
@@ -0,0 +1,77 @@
+# Title: Password Key Derivation
+
+## Abstract
+
+This design doc describes the way cryptographic encryption keys will be derived from user provided passwords in a way that resists offline attacks and low entropy passwords.
+
+## Background
+
+A unique encryption key exists for every path inside of a bucket, and they are designed in a way that allows for deriving a key from a parent and some path component. For example, if you had the key for `sj://bucket/path1/path2` then you could derive the key for `sj://bucket/path1/path2/path3`.
+
+That leaves open the question for how a root key is derived. Some requirements on that root key derivation include
+
+1. It's based on the user's choice (for example, a password).
+2. If the user inputs the same password on different machines, the same root key should be derived.
+
+These requirements allow users to be in full control of their encryption, and don't require users safely transporting high entropy (hard to remember) secrets to bootstrap new uplinks.
+
+This design accomodates more requirements that allow for additional features:
+
+3. A root key can be derived for any path in a bucket.
+4. Buckets root keys should not depend on the name of the bucket.
+5. A table of keys for low entropy passwords should not be possible. In other words, an attacker with knowledge of the algorithm should not be able to use a dictionary of common passwords and pre-compute what keys to check in the event of a data breach.
+6. Users do not have to enter a password for every bucket they create: they can have a default password that is used for every bucket unless otherwise specified.
+
+The third requirement allows having multiple encrypted domains to exist within a single bucket, allowing delegation of encryption. The fourth requirement allows renaming buckets without having to re-encrypt all of the paths inside of a bucket. The fifth requirement enhances security if a satellite breach happens. This is usually accomplished with [salting](https://en.wikipedia.org/wiki/Salt_(cryptography)). The sixth requirement allows providing a finite amount of information to a third party that allows them access to the keys for any bucket created with the default password in the future.
+
+## Design
+
+Satellites will add a `salt` field to bucket metadata that lives with the path encryption information that already exists and that we already request. Satellites will also store a `salt` that lives with the project metadata. The salts will be 32 random, stable bytes.
+
+A default password can be chosen for an api key. The following algorithm is used to derive the default password:
+
+```
+saltEntropy     = getProjectSalt(apiKey)
+mixedSalt       = hmac(hash=sha256, secret=userPassword, data=saltEntropy)
+defaultPassword = argon2id(salt=mixedSalt, password=userPassword)
+```
+
+The following algorithm is used to deriving a root key from an api key, password, bucket name and optional encrypted path:
```

Comment:
The term "root key" is quite overloaded in this document. Is the "root key", referred here, the key that we derive for a specific bucket? And then we derive keys for subpath from this key.

If my understanding is correct, can we name them "bucket encryption key" and "path encryption key"?

Reply:
I'll add a section that explicitly defines "root key", but it may be the key used at the root of the bucket, or it may be a key used deeper in the bucket that would not match what would be the key from an earlier root.

For example, imagine one is making a dropbox style product and they want users in control of their encryption. There may be a `userdata` bucket with some root key, and it would contain paths like `sj://userdata/user1` and `sj://userdata/user2`. Those users could choose their own password and have a new root key at those paths, delegating encryption.

In other words, it's allowing the s<sub>0</sub> key (root key) to exist at a path other than the empty path.